src/main/java/com/github/splendor_mobile_game/game/model/User.java

```diff
@@ -90,24 +92,30 @@ public boolean takeThreeTokens(TokenType type1, TokenType type2, TokenType type3
     //method for buying cards
     public void buyCard(Card card) throws NotEnoughTokensException {
 
+        int goldTokensUsed = 0;
+
         for(Map.Entry<TokenType, Integer> set : this.tokens.entrySet()) {
             if(set.getKey() == TokenType.GOLD_JOKER) continue;
-            if(set.getValue() < card.getCost(set.getKey())) throw new NotEnoughTokensException("You don't have enough tokens to buy this card");
+            int tokensPlusBonuses = set.getValue() + this.cardBonuses.get(set.getKey());
+            if(tokensPlusBonuses < card.getCost(set.getKey())) {
+                int missingTokens = card.getCost(set.getKey()) - tokensPlusBonuses;
+                if((this.getTokenCount(TokenType.GOLD_JOKER) - goldTokensUsed) >= missingTokens) {
+                    goldTokensUsed += missingTokens;
+                } else {
+                    throw new NotEnoughTokensException("You don't have enough tokens to buy this card");
+                }
+            }
         }
 
         this.tokens.forEach((k,v) -> {
             if(k != TokenType.GOLD_JOKER) {
-                this.tokens.put(k, v - card.getCost(k));
+                int neededTokens = card.getCost(k) - this.cardBonuses.get(k);
+                int changedValue = v > neededTokens ? v - neededTokens : 0;
```

Comment:
So I see, here we paying with our normal tokens, where do we give out our gold tokens if they're gonna by used?

Reply:
In goldTokensUsed variable i count how many tokens are needed, and in line 121 I take this amount of gold tokens from user :)

Reply:
*line 119